packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js

```diff
@@ -877,6 +877,60 @@ QUnit.module('MessageList', () => {
                 assert.strictEqual($lastGroup.length, 1, 'only one message group has expected class');
             });
         });
+
+        QUnit.test('messageTemplate should set bubble content on init', function(assert) {
+            const messageTemplate = ({ text }, container) => {
+                $('<h1>').text(`${text}${text}`).appendTo(container);
+            };
+
+            this.reinit({
+                items: [{ text: 'CustomText' }],
+                messageTemplate,
+            });
+
+            const $bubble = this.getBubbles();
+
+            assert.strictEqual($bubble.text(), 'CustomTextCustomText');
+        });
+
+        QUnit.test('messageTemplate should set bubble content at runtime', function(assert) {
+            const messageTemplate = ({ text }, container) => {
+                $('<h1>').text(`${text}${text}`).appendTo(container);
+            };
+
+            this.reinit({
+                items: [{ text: 'CustomText' }]
+            });
+
+            this.instance.option('messageTemplate', messageTemplate);
+
+            const $bubble = this.getBubbles();
+
+            assert.strictEqual($bubble.text(), 'CustomTextCustomText');
+        });
+
+        QUnit.test('messageTemplate function should have correct parameters', function(assert) {
+            assert.expect(3);
+
+            const timestamp = 1234567;
+            const text = 'message text';
+            const author = { name: 'UserName', id: 'UserID' };
+
+            const messageTemplate = (data) => {
+                assert.deepEqual(data.author, author, 'author parameter is passed');
+                assert.strictEqual(data.timestamp, timestamp, 'timestamp parameter is passed');
+                assert.strictEqual(data.text, text, 'text parameter is passed');
+            };
+
+            this.reinit({
+                items: [{
+                    timestamp,
+                    text,
+                    author,
+                }],
+                messageTemplate,
+            });
+        });
```

Comment:
Okay, now I see you copy-paste these 3 tests across all component (chat/messageList/messageGroup/messageBubble). What's the idea? Chat test cannot be green if one of these tests is red. Cannot we use our general approach with testing the functionality on level where we implemented it + test that necessary properties are passed to subcomponents correctly?
P.s. I don't mind about leaving there "real scenario test" on Chat level, but I'm bothered we repeat it for the each component + don't test subcomponents' new API. 
P.s.s. If you cannot write good tests, maybe that's a hint that solution is overcomplicated?

Reply:
3c18eb83217a111cc7e4af6de5eef1636d87ee02